src/sb16.js

```diff
@@ -0,0 +1,1620 @@
+{
+    // No more data to transfer.
+    if(!this.dma_bytes_left) return;
+
+    // DAC has enough samples buffered for now.
+    // Don't transfer too much too early, or else the DMA counters will not
+    // accurately reflect the amount of audio that has already been
+    // played back by the Web Audio API.
+    if(this.dac_buffers[0].length > this.dac_process_samples * 2) return;
+
+    // Do not transfer if paused.
+    if(this.cpu_paused || this.dma_paused) return;
+
+    dbg_log("dma transfering next block", LOG_SB16);
+
+    var size = Math.min(this.dma_bytes_left, this.dma_bytes_block);
+    var samples = Math.floor(size / this.bytes_per_sample);
+
+    this.dma.do_write(this.dma_syncbuffer, 0, size, this.dma_channel, (error) =>
+    {
+        dbg_log("dma block transfer " + (error ? "unsuccessful" : "successful"), LOG_SB16);
+        if(error) return;
+
+        this.dma_to_dac(samples);
+        this.dma_bytes_left -= size;
+
+        if(!this.dma_bytes_left)
+        {
+            // Completed requested transfer of given size.
+            this.raise_irq(this.dma_irq);
+
+            if(this.dma_autoinit)
+            {
+                // Restart the transfer.
+                this.dma_bytes_left = this.dma_bytes_count;
+            }
+        }
+
+        // Keep transfering until dac_buffer contains enough data.
+        setTimeout(() => { this.dma_transfer_next(); }, 0);
+    });
+};
+
+SB16.prototype.dma_to_dac = function(sample_count)
+{
+    var amplitude = this.dsp_16bit? 32767.5 : 127.5;
+    var offset = this.dsp_signed? 0 : -1;
+    var repeats = (this.dsp_stereo? 1 : 2) * this.dac_rate_ratio;
+
+    var buffer;
+    if(this.dsp_16bit)
+    {
+        buffer = this.dsp_signed ?
+            this.dma_buffer_int16 :
+            this.dma_buffer_uint16;
+    }
+    else
+    {
+        buffer = this.dsp_signed ?
+            this.dma_buffer_int8 :
+            this.dma_buffer_uint8;
+    }
+
+    var channel = 0;
+    for(var i = 0; i < sample_count; i++)
+    {
+        for(var j = 0; j < repeats; j++)
+        {
+            this.dac_buffers[channel].push(audio_normalize(buffer[i], amplitude, offset));
+            channel ^= 1;
+        }
+    }
+};
+
+SB16.prototype.audio_send = function(size)
+{
+    this.dac_process_samples = size;
+
+    if(this.cpu_paused || this.dma_paused)
+    {
+        // Silence when paused.
+        var silence = new Float32Array(size);
+        this.bus.send("speaker-update-data", [silence, silence]);
+        return;
+    }
+
+    if(this.dac_buffers[0].length && this.dac_buffers[0].length < this.dac_process_samples * 2)
+    {
+        dbg_log("dac_buffer contains only " +
+            (this.dac_buffers[0].length / 2) +
+            " samples out of " + this.dac_process_samples + " needed", LOG_SB16);
+    }
+
+    this.bus.send("speaker-update-data",
+    [
+        this.dac_buffers[0].shift_block(size),
+        this.dac_buffers[1].shift_block(size),
```

Comment:
The underlying buffer of these should be passed as the third argument to avoid another copy.